apps/taxes/src/modules/avatax/calculate-taxes/avatax-calculate-taxes-payload.service.ts

```diff
@@ -0,0 +1,26 @@
+import { AuthData } from "@saleor/app-sdk/APL";
+import { TaxBaseFragment } from "../../../../generated/graphql";
+import { CreateTransactionArgs } from "../avatax-client";
+import { AvataxConfig } from "../avatax-connection-schema";
+import { AvataxTaxCodeMatchesService } from "../tax-code/avatax-tax-code-matches.service";
+import { AvataxCalculateTaxesPayloadTransformer } from "./avatax-calculate-taxes-payload-transformer";
+
+export class AvataxCalculateTaxesPayloadService {
+  constructor(private ctx: AuthData) {}
```

Comment:
I'm not convinced that you should pass AuthData to so many classes. 

Authdata is an abstractin on endpoint/controller level. In the endpoint you create a Client from it. Your services need client to perform operations, not checking if user is authorized

Reply:
There are two things here:
- `Client` - I create it somewhere between the webhook and the metadata manager call, so you can make an argument I can just create it on the top and pass it where it's needed, but...
- `saleorApiUrl` (which is a part of the `AuthData`) - I need to pass it all the way down to the metadata manager.

So it was more convenient for me to pass the entire `AuthData` since in one place, I need `token` and `appId`, and in another (basically on the bottom of the call stack) I need `saleorApiUrl`. 

Best I can do is pass `client` and `saleorApiUrl`

Reply:
The original issue comes from `MetadataManager` requiring a domain in the method arguments instead of accepting it in the constructor.

If it did, I could create a metadata manager at the top (where I have the entire `AuthData`, along with `saleorApiUrl`), and just pass it down where it's needed along with the client.

But because it requires `saleorApiUrl` in the method calls, I need to pass it to the bottom of the call stack (where I call methods from a metadata manager).

Do you see a solution here? I can create a wrapper class like:

```
class DomainMetadataManager extends EncryptedMetadataManager {
  private domain: string;

  constructor(client: Client, appId: string, domain: string) {
    super({
      encryptionKey: process.env.SECRET_KEY!,
      fetchMetadata: () => fetchAllMetadata(client),
      mutateMetadata: (metadata) => mutateMetadata(client, metadata, appId),
    });
    this.domain = domain;
  }

  async get(key: string) {
    return super.get(key, this.domain);
  }

  async set(settings: DomainSettingsValue[] | DomainSettingsValue) {
    if (Array.isArray(settings)) {
      return super.set(
        settings.map((s) => ({
          ...s,
          domain: this.domain,
        }))
      );
    }

    return super.set({
      ...settings,
      domain: this.domain,
    });
  }
}

```

or we can just modify the original metadata manager, unless it was meant to be able to modify more than one domain under one class.

Reply:
It's not only used through the trpc router, it's also needed when called from webhooks